# Problem Statement
- This Dataset Contain the information about the bank customer that will take loan.
- Our goal is build a model that will detect weather the customer is eligible for loan or not.
- Data should be present in 2 different files.
- Both conatin same data about bank.
- Both files also contain `-99999` values.
- `-99999` are the missing values.

# Import Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')

# Read Data
1. Data is place another folder now find the path and read data.

In [2]:
file_path=[]
for i in os.listdir("Data"):
    file_path.append(f"{os.path.join(os.getcwd())}\Data\{i}")
    print(f"{os.path.join(os.getcwd())}\Data\{i}")

C:\Users\Laptop House\Weekly Task\CreditRisk\Data\case_study1.xlsx
C:\Users\Laptop House\Weekly Task\CreditRisk\Data\case_study2.xlsx


In [46]:
df1=pd.read_excel(file_path[0])
df2=pd.read_excel(file_path[1])

# Data Preprocessing
1. Top records
2. Shape of data
3. Check columns
4. Merge based on `common column`
5. Check Datatyes
6. Null values
7. Duplicates
8. Unique Values
9. Statistical Summary

# 1.Top Records

In [47]:
df1.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
48268       48269         1              0              1   
407           408         1              1              0   
19207       19208         1              1              0   
7458         7459         1              0              1   
48424       48425         2              1              1   
48028       48029         2              1              1   
28267       28268         1              0              1   
2748         2749         6              2              4   
49905       49906         6              2              4   
48917       48918         7              4              3   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
48268                    1                  0            1.000   
407                      0                  0            0.000   
19207                    0                  1            0.000   
7458                     0                  0            0.000   
48424                    1                  1            0.500   
48028                    0                  0            0.000   
28267                    0                  0            0.000   
2748                     0                  0            0.000   
49905                    2                  1            0.333   
48917                    0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  Total_TL_opened_L12M  \
48268              0.000          1.000          0.000                     1   
407                0.000          0.000          1.000                     0   
19207              1.000          0.000          1.000                     0   
7458               0.000          1.000          0.000                     1   
48424              0.500          0.500          0.500                     1   
48028              0.000          0.500          0.500                     1   
28267              0.000          1.000          0.000                     0   
2748               0.000          0.667          0.333                     1   
49905              0.167          0.667          0.333                     5   
48917              0.000          0.429          0.571                     0   

       Tot_TL_closed_L12M  pct_tl_open_L12M  pct_tl_closed_L12M  \
48268                   0             1.000               0.000   
407                     1             0.000               1.000   
19207                   1             0.000               1.000   
7458                    0             1.000               0.000   
48424                   1             0.500               0.500   
48028                   0             0.500               0.000   
28267                   0             0.000               0.000   
2748                    0             0.167               0.000   
49905                   2             0.833               0.333   
48917                   0             0.000               0.000   

       Tot_Missed_Pmnt  Auto_TL  CC_TL  Consumer_TL  Gold_TL  Home_TL  PL_TL  \
48268                0        0      0            0        1        0      0   
407                  0        0      0            1        0        0      0   
19207                0        1      0            0        0        0      0   
7458                 0        0      0            1        0        0      0   
48424                0        0      0            2        0        0      0   
48028                0        2      0            0        0        0      0   
28267                0        0      0            0        0        0      0   
2748                 0        0      1            1        1        0      1   
49905                4        0      0            0        6        0      0   
48917                0        0      0            4        0        0      1   

       Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  Age_Newest_TL  
48268           1             0   

In [48]:
df2.sample(10)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
1574         1575                         76                       -99999   
4159         4160                         72                            2   
16455       16456                       2112                       -99999   
5299         5300                         67                       -99999   
35834       35835                        780                       -99999   
4924         4925                         59                       -99999   
37305       37306                        141                       -99999   
28860       28861                        123                           11   
2789         2790                        103                       -99999   
40513       40514                        286                            8   

       time_since_recent_deliquency  num_times_delinquent  \
1574                         -99999                     0   
4159                              2                     1   
16455                        -99999                     0   
5299                         -99999                     0   
35834                        -99999                     0   
4924                         -99999                     0   
37305                        -99999                     0   
28860                             5                     5   
2789                         -99999                     0   
40513                             8                     2   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
1574                  -99999                          0               0   
4159                       9                          9               1   
16455                 -99999                          0               0   
5299                  -99999                          0               0   
35834                 -99999                          0               0   
4924                  -99999                          0               0   
37305                 -99999                          0               0   
28860                      4                          4               1   
2789                  -99999                          0               0   
40513                     22                         22               0   

       num_deliq_12mts  num_deliq_6_12mts  max_deliq_6mts  max_deliq_12mts  \
1574                 0                  0               0                0   
4159                 1                  0               9                9   
16455                0                  0          -99999           -99999   
5299                 0                  0               0                0   
35834                0                  0          -99999           -99999   
4924                 0                  0               0                0   
37305                0                  0          -99999           -99999   
28860                5                  4               4                4   
2789                 0                  0               0                0   
40513                2                  2               0               22   

       num_times_30p_dpd  num_times_60p_dpd  num_std  num_std_6mts  \
1574                   0                  0       11             3   
4159                   0                  0        0             0   
16455                  0                  0        0             0   
5299                   0                  0        0             0   
35834                  0                  0        0             0   
4924                   0                  0        0             0   
37305                  0                  0       41             5   
28860                  0                  0        0             0   
2789                   0                  0        0             0   
40513                  0                  0        0             0   

       num_std_12mts  num_sub  

# Conclussion
1. we can see that in `df2` most of the columns contain `-99999` values.
2. These `-99999` are basically `Missing value`.

# 2.shape

In [49]:
print("Data 1 Shape: ",df1.shape)
print("Data 2 Shape: ",df2.shape)

Data 1 Shape:  (51336, 26)
Data 2 Shape:  (51336, 62)


# Conclussion 
1. Data1 contain `51336` rows and `26` Column
2. Data2 contain `51336` rows and `62` Column

# 3.Columns

In [50]:
print(df1.columns)
print(df2.columns)

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL'],
      dtype='object')
Index(['PROSPECTID', 'time_since_recent_payment',
       'time_since_first_deliquency', 'time_since_recent_deliquency',
       'num_times_delinquent', 'max_delinquency_level',
       'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts',
       'num_deliq_6_12mts', 'max_deliq_6mts', 'max_deliq_12mts',
       'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std', 'num_std_6mts',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_6mts', 'num_dbt_

# Conclussion
1. See that both `dataframe` one common column `PROSPECTID`.
2. We can `merge` them first and then perform other data preprocessing.

# 4.Merge `df1` anf `df2`

In [51]:
df1.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  Total_TL_opened_L12M  Tot_TL_closed_L12M  pct_tl_open_L12M  \
0            0.8                     0                   0               0.0   
1            0.0                     1                   0               1.0   

   pct_tl_closed_L12M  Tot_Missed_Pmnt  Auto_TL  CC_TL  Consumer_TL  Gold_TL  \
0                 0.0                0        0      0            0        1   
1                 0.0                0        0      0            1        0   

   Home_TL  PL_TL  Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  \
0        0      4           1             4         0             72   
1        0      0           0             1         0              7   

   Age_Newest_TL  
0             18  
1              7

In [52]:
df2.head(2)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
0           1                        549                           35   
1           2                         47                       -99999   

   time_since_recent_deliquency  num_times_delinquent  max_delinquency_level  \
0                            15                    11                     29   
1                        -99999                     0                 -99999   

   max_recent_level_of_deliq  num_deliq_6mts  num_deliq_12mts  \
0                         29               0                0   
1                          0               0                0   

   num_deliq_6_12mts  max_deliq_6mts  max_deliq_12mts  num_times_30p_dpd  \
0                  0          -99999           -99999                  0   
1                  0               0                0                  0   

   num_times_60p_dpd  num_std  num_std_6mts  num_std_12mts  num_sub  \
0                  0       21             5             11        0   
1                  0        0             0              0        0   

   num_sub_6mts  num_sub_12mts  num_dbt  num_dbt_6mts  num_dbt_12mts  num_lss  \
0             0              0        0             0              0        0   
1             0              0        0             0              0        0   

   num_lss_6mts  num_lss_12mts  recent_level_of_deliq  tot_enq  CC_enq  \
0             0              0                     29        6       0   
1             0              0                      0        1       0   

   CC_enq_L6m  CC_enq_L12m  PL_enq  PL_enq_L6m  PL_enq_L12m  \
0           0            0       6           0            0   
1           0            0       0           0            0   

   time_since_recent_enq  enq_L12m  enq_L6m  enq_L3m MARITALSTATUS EDUCATION  \
0                    566         0        0        0       Married      12TH   
1                    209         1        0        0        Single  GRADUATE   

   AGE GENDER  NETMONTHLYINCOME  Time_With_Curr_Empr  pct_of_active_TLs_ever  \
0   48      M             51000                  114                     0.2   
1   23      F             19000                   50                     1.0   

   pct_opened_TLs_L6m_of_L12m  pct_currentBal_all_TL  CC_utilization  CC_Flag  \
0                         0.0                  0.798        -99999.0        0   
1                         0.0                  0.370        -99999.0        0   

   PL_utilization  PL_Flag  pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  \
0           0.798        1                     0.0                     0.0   
1      -99999.000        0                     0.0                     0.0   

   pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  \
0                     0.0                     0.0                    13.333   
1                     0.0                     0.0                     0.860   

   HL_Flag  GL_Flag last_prod_enq2 first_prod_enq2  Credit_Score Approved_Flag  
0        1        0             PL              PL           696            P2  
1        0        0   ConsumerLoan    ConsumerLoan           685            P2

In [97]:
final_df=df1.merge(df2,how='inner',on='PROSPECTID')
final_df.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  Total_TL_opened_L12M  Tot_TL_closed_L12M  pct_tl_open_L12M  \
0            0.8                     0                   0               0.0   
1            0.0                     1                   0               1.0   

   pct_tl_closed_L12M  Tot_Missed_Pmnt  Auto_TL  CC_TL  Consumer_TL  Gold_TL  \
0                 0.0                0        0      0            0        1   
1                 0.0                0        0      0            1        0   

   Home_TL  PL_TL  Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  \
0        0      4           1             4         0             72   
1        0      0           0             1         0              7   

   Age_Newest_TL  time_since_recent_payment  time_since_first_deliquency  \
0             18                        549                           35   
1              7                         47                       -99999   

   time_since_recent_deliquency  num_times_delinquent  max_delinquency_level  \
0                            15                    11                     29   
1                        -99999                     0                 -99999   

   max_recent_level_of_deliq  num_deliq_6mts  num_deliq_12mts  \
0                         29               0                0   
1                          0               0                0   

   num_deliq_6_12mts  max_deliq_6mts  max_deliq_12mts  num_times_30p_dpd  \
0                  0          -99999           -99999                  0   
1                  0               0                0                  0   

   num_times_60p_dpd  num_std  num_std_6mts  num_std_12mts  num_sub  \
0                  0       21             5             11        0   
1                  0        0             0              0        0   

   num_sub_6mts  num_sub_12mts  num_dbt  num_dbt_6mts  num_dbt_12mts  num_lss  \
0             0              0        0             0              0        0   
1             0              0        0             0              0        0   

   num_lss_6mts  num_lss_12mts  recent_level_of_deliq  tot_enq  CC_enq  \
0             0              0                     29        6       0   
1             0              0                      0        1       0   

   CC_enq_L6m  CC_enq_L12m  PL_enq  PL_enq_L6m  PL_enq_L12m  \
0           0            0       6           0            0   
1           0            0       0           0            0   

   time_since_recent_enq  enq_L12m  enq_L6m  enq_L3m MARITALSTATUS EDUCATION  \
0                    566         0        0        0       Married      12TH   
1                    209         1        0        0        Single  GRADUATE   

   AGE GENDER  NETMONTHLYINCOME  Time_With_Curr_Empr  pct_of_active_TLs_ever  \
0   48      M             51000                  114                     0.2   
1   23      F             19000                   50                     1.0   

   pct_opened_TLs_L6m_of_L12m  pct_currentBal_all_TL  CC_utilization  CC_Flag  \
0                         0.0                  0.798        -99999.0        0   
1                         0.0                  0.370        -99999.0        0   

   PL_utilization  PL_Flag  pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  \
0           0.798        1                     0.0                     0.0   
1      -99999.000        0                     0.0                     0.0   

   pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  \
0                     0.0            

# Conclussion
1. we have a `final df` now we can further work on this `final_df`

# 5. Datatypes

In [98]:
data_types=final_df.dtypes.to_dict()

In [99]:
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns
pd.DataFrame(list(data_types.items()),columns=['Columns','Datatypes']).head(87)

Columns Datatypes
0                     PROSPECTID     int64
1                       Total_TL     int64
2                  Tot_Closed_TL     int64
3                  Tot_Active_TL     int64
4            Total_TL_opened_L6M     int64
5              Tot_TL_closed_L6M     int64
6                pct_tl_open_L6M   float64
7              pct_tl_closed_L6M   float64
8                  pct_active_tl   float64
9                  pct_closed_tl   float64
10          Total_TL_opened_L12M     int64
11            Tot_TL_closed_L12M     int64
12              pct_tl_open_L12M   float64
13            pct_tl_closed_L12M   float64
14               Tot_Missed_Pmnt     int64
15                       Auto_TL     int64
16                         CC_TL     int64
17                   Consumer_TL     int64
18                       Gold_TL     int64
19                       Home_TL     int64
20                         PL_TL     int64
21                    Secured_TL     int64
22                  Unsecured_TL     int64
23                      Other_TL     int64
24                 Age_Oldest_TL     int64
25                 Age_Newest_TL     int64
26     time_since_recent_payment     int64
27   time_since_first_deliquency     int64
28  time_since_recent_deliquency     int64
29          num_times_delinquent     int64
30         max_delinquency_level     int64
31     max_recent_level_of_deliq     int64
32                num_deliq_6mts     int64
33               num_deliq_12mts     int64
34             num_deliq_6_12mts     int64
35                max_deliq_6mts     int64
36               max_deliq_12mts     int64
37             num_times_30p_dpd     int64
38             num_times_60p_dpd     int64
39                       num_std     int64
40                  num_std_6mts     int64
41                 num_std_12mts     int64
42                       num_sub     int64
43                  num_sub_6mts     int64
44                 num_sub_12mts     int64
45                       num_dbt     int64
46                  num_dbt_6mts     int64
47                 num_dbt_12mts     int64
48                       num_lss     int64
49                  num_lss_6mts     int64
50                 num_lss_12mts     int64
51         recent_level_of_deliq     int64
52                       tot_enq     int64
53                        CC_enq     int64
54                    CC_enq_L6m     int64
55                   CC_enq_L12m     int64
56                        PL_enq     int64
57                    PL_enq_L6m     int64
58                   PL_enq_L12m     int64
59         time_since_recent_enq     int64
60                      enq_L12m     int64
61                       enq_L6m     int64
62                       enq_L3m     int64
63                 MARITALSTATUS    object
64                     EDUCATION    object
65                           AGE     int64
66                        GENDER    object
67              NETMONTHLYINCOME     int64
68           Time_With_Curr_Empr     int64
69        pct_of_active_TLs_ever   float64
70    pct_opened_TLs_L6m_of_L12m   float64
71         pct_currentBal_all_TL   float64
72                CC_utilization   float64
73                       CC_Flag     int64
74                PL_utilization   float64
75                       PL_Flag     int64
76        pct_PL_enq_L6m_of_L12m   float64
77        pct_CC_enq_L6m_of_L12m   float64
78        pct_PL_enq_L6m_of_ever   float64
79        pct_CC_enq_L6m_of_ever   float64
80      max_unsec_exposure_inPct   float64
81                       HL_Flag     int64
82                       GL_Flag     int64
83                last_prod_enq2    object
84               first_prod_enq2    object
85                  Credit_Score     int64
86                 Approved_Flag    object

In [100]:
# object colums need =['material_status','Education','gender','last_prod_loan','first_prod_loan','Approved_flag']
# int columns need =[All except object col]

In [101]:
final_df.select_dtypes('object').columns

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

In [102]:
final_df.select_dtypes('number').columns

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL',
       'time_since_recent_payment', 'time_since_first_deliquency',
       'time_since_recent_deliquency', 'num_times_delinquent',
       'max_delinquency_level', 'max_recent_level_of_deliq', 'num_deliq_6mts',
       'num_deliq_12mts', 'num_deliq_6_12mts', 'max_deliq_6mts',
       'max_deliq_12mts', 'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std',
       'num_std_6mts', 'num_std_12mts', 'num_sub', 'num_sub_6mts',
       'num_sub_12mts', 'num_dbt', 'num_dbt_6mts', 'num_dbt_12mts', 'num_lss',
       'num_lss_6mts', 'n

# Conclussion
1. `Datatypes` should be in correct format.

# 6. Null Values
1. First we can make the copy of dataframe

In [103]:
copy_df=final_df.copy()

In [104]:
final_df.isnull().sum()

PROSPECTID                      0
Total_TL                        0
Tot_Closed_TL                   0
Tot_Active_TL                   0
Total_TL_opened_L6M             0
Tot_TL_closed_L6M               0
pct_tl_open_L6M                 0
pct_tl_closed_L6M               0
pct_active_tl                   0
pct_closed_tl                   0
Total_TL_opened_L12M            0
Tot_TL_closed_L12M              0
pct_tl_open_L12M                0
pct_tl_closed_L12M              0
Tot_Missed_Pmnt                 0
Auto_TL                         0
CC_TL                           0
Consumer_TL                     0
Gold_TL                         0
Home_TL                         0
PL_TL                           0
Secured_TL                      0
Unsecured_TL                    0
Other_TL                        0
Age_Oldest_TL                   0
Age_Newest_TL                   0
time_since_recent_payment       0
time_since_first_deliquency     0
time_since_recent_deliquency    0
num_times_deli

1. **We can't see missing values but in above we say that `-99999` shows the missing values**
2. **Check if the ratio of `-99999` is greater in the specific column remove the columns**

# Handling Missing Values

In [105]:
final_df[final_df['time_since_recent_payment']==-99999].shape[0]

4291

In [106]:
final_df.shape

(51336, 87)

1. **We have `51336` records we can set the treshold `10000` if any columns contain `-99999` more then 10000 we can drop the columns.** 
2. **Also check the ratio of containg `-99999`.**

In [107]:
col_to_drop=[]
for i in final_df.columns:
    temp=final_df[final_df[i]==-99999].shape[0]
    ratio= round(temp/len(final_df),3)*100
    if(temp>10000 or temp2>10000 ):
        print(f"{i} -> -99999 value is -> {temp} and ratio is {ratio}%")
        col_to_drop.append(i)

time_since_first_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
time_since_recent_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
max_delinquency_level -> -99999 value is -> 35949 and ratio is 70.0%
max_deliq_6mts -> -99999 value is -> 12890 and ratio is 25.1%
max_deliq_12mts -> -99999 value is -> 10832 and ratio is 21.099999999999998%
CC_utilization -> -99999 value is -> 47636 and ratio is 92.80000000000001%
PL_utilization -> -99999 value is -> 44435 and ratio is 86.6%
max_unsec_exposure_inPct -> -99999 value is -> 23178 and ratio is 45.1%


# Conclussion
1. `time_since_first_deliquency` contain `70.0%` missing data.
2. `time_since_recent_deliquency` contain `70.0%` missing data.
3. `max_delinquency_level` contain `70.0%` missing data.
4. `max_deliq_6mts` contain `25.1%` missing data.
5. `max_deliq_12mts` contain  `21.09% missing data.
6. `CC_utilization` conrain `92.8%` missing data.
7. `PL_utilization` contain `86.6%` missing data.
8. `max_unsec_exposure_inPct` contain  `45.1%` missing data.
- **We will drop these columns**

### Drop Columns 

In [108]:
final_df.drop(columns=col_to_drop,inplace=True)

In [109]:
final_df.shape

(51336, 79)

In [110]:
final_df.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
43754       43755         5              3              2   
33737       33738         2              0              2   
2751         2752         1              0              1   
14242       14243         6              3              3   
34523       34524        15             10              5   
2888         2889         1              0              1   
14194       14195         2              0              2   
45446       45447         1              0              1   
17545       17546         6              1              5   
8401         8402         3              1              2   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
43754                    0                  0            0.000   
33737                    0                  0            0.000   
2751                     0                  0            0.000   
14242                    1                  1            0.167   
34523                    0                  1            0.000   
2888                     0                  0            0.000   
14194                    2                  0            1.000   
45446                    0                  0            0.000   
17545                    3                  0            0.500   
8401                     0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  Total_TL_opened_L12M  \
43754              0.000          0.400          0.600                     1   
33737              0.000          1.000          0.000                     1   
2751               0.000          1.000          0.000                     1   
14242              0.167          0.500          0.500                     2   
34523              0.067          0.333          0.667                     2   
2888               0.000          1.000          0.000                     0   
14194              0.000          1.000          0.000                     2   
45446              0.000          1.000          0.000                     0   
17545              0.000          0.833          0.167                     3   
8401               0.000          0.667          0.333                     2   

       Tot_TL_closed_L12M  pct_tl_open_L12M  pct_tl_closed_L12M  \
43754                   0             0.200               0.000   
33737                   0             0.500               0.000   
2751                    0             1.000               0.000   
14242                   1             0.333               0.167   
34523                   3             0.133               0.200   
2888                    0             0.000               0.000   
14194                   0             1.000               0.000   
45446                   0             0.000               0.000   
17545                   0             0.500               0.000   
8401                    1             0.667               0.333   

       Tot_Missed_Pmnt  Auto_TL  CC_TL  Consumer_TL  Gold_TL  Home_TL  PL_TL  \
43754                0        2      1            0        1        0      0   
33737                1        0      0            1        0        0      0   
2751                 0        1      0            0        0        0      0   
14242                1        3      0            2        0        0      1   
34523                0        3      1            6        3        0      1   
2888                 0        0      0            0        0        0      1   
14194                0        0      0            1        0        0      1   
45446                0        0      0            0        0        0      0   
17545                0        0      1            0        4        0      1   
8401                 1        1      0            0        2        0      0   

       Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  Age_Newest_TL  \
43754           4             1  

# Conclussion
- After removing columns which contain `-99999` values more then `10000` shape of the dataframe is `51336` rows and `79` columns.
- we can remove only those columns which contain `-99999` values more then `10000` but remember that `-99999` are also present in those columns which contain  `-99999` values less then `10000`.
# Point to remember
- we have 2 option for `handling the remaing missing values`.
    ## Remove Missing Values
    - first we can remove those value.
        - we can remove the missing values if we maintain the balance of data we can remove them.
    ## Fill Missing Values
    - Fill these values
        - Filling the missing values is a cretical thing but if we can loss lot of data after removing the missing values then we pick the filling missing values.

- **Try to remove rows that contain `-99999` value**

In [111]:
for i in final_df.columns:
    print(final_df[~(final_df[i]==-99999)].shape[0])
    final_df=final_df[~(final_df[i]==-99999)]

51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51296
51296
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064


# Conclussion
- Check the shape after removing.
- Also check the ration before removing and after removing

In [112]:
print(f"Shape of the data after removing -9999 values is: {final_df.shape}")

Shape of the data after removing -9999 values is: (42064, 79)


In [117]:
len(final_df)/len(copy_df)*100

81.93860059217702

# Conclussion
- we see that after removing the rows that contain `-99999` values we can maintain `80%` data.
- Now we sholud pick remove missing value and we will remove them.